### ISSUE TITLE
Create more IDE friendly API

### ISSUE THREAD

**Issue by StokeMasterJack** [ktorio/ktor#131]
One reason I use kotlin over, say, nodejs is that I get nice IDE autocompletion, so that i don't need api docs. But when you get too fancy with the api, these benefits go away. For example:

`val map = call.request.receive<ValuesMap>()`

There is no obvious way to determine what type parameters are valid for receive without searching through the source code. I would prefer simple typed methods for the 90% cases:

```
val content:Map<String,List<String>> call.request.content.asParamMap()
val content:String call.request.content.asString()
val content:Byte[] call.request.content.asByteArray()
val content:ReadChannel call.request.content.asReadChannel()
val content:InputStream call.request.content.asInputStream()
val content:MultiPartData call.request.content.asMultiPartData()
```
And also:

`val x:String = call.request["x"] //works like servletRequest.getParameter("x")`

Also, if one calls `asParamMap` on a request where `contentType != FormUrlEncoded`, then do this:

`throw WrongContentTypeException("asParamMap can only be called for contentType = FormUrlEncoded")`

rather than this:

`return ValuesMap.Empty`

IOW "fail fast"

**Comment by StokeMasterJack** [comment]
I ended up writing a blog post about this:

https://medium.com/@daveford/kotlin-dont-repeat-the-same-mistakes-java-made-7867e9d25b99

**Comment by orangy** [comment]
Addressed this partially in 0.4.0 with new `call.receiveTTT()` functions

**Comment by orangy** [comment]
I think 0.4 addressed all of these issues, no? Should we close it or are there any other issues you think needs to be addressed?
